Add guide for AsyncChatEvent

Leguan16 opened the pull request:
This PR adds a guide for the `AsyncChatEvent` as people seem to strugle with it. Especially with the renderer.

Machine-Maker commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,154 @@
+
+# Chat Events
+The chat event has evolved a couple of times over the years. 
+This guide will explain how to properly use the new `AsyncChatEvent` and its `Renderer`.
```
```suggestion
This guide will explain how to properly use the new `AsyncChatEvent` and its `ChatRenderer`.
```

Machine-Maker commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,154 @@
+
+## Understanding the Renderer
+
+Before we can start using the new chat event, we need to understand how the new renderer works.
+The renderer is Paper's way of allowing plugins to modify the chat message before it is sent to the player.
+This is done by using the `ChatRenderer` interface with its `render` method.
```
I would maybe also mention somewhere that it is an improved replacement for the deprecated AsyncPlayerChatEvent#setFormat so authors who know of that can make a connection between the two.

Machine-Maker commented on docs/paper/dev/api/event-api/chat-event.md:
````diff
@@ -0,0 +1,154 @@
+}
+```
+
+- The `render` method is called when a chat message is sent to the player.
+- The `source` parameter is the player that sent the message.
+- The `sourceDisplayName` parameter is the display name of the player that sent the message.
+- The `message` parameter is the message that was sent.
+- The `viewer` parameter is the player that is receiving the message.
+
+Now that we understand how the renderer works, we can start using it.
````
I think before this happens, it needs to be made clear that the render method is something that plugins implement, they generally do not call it themselves. And they don't supply each of those arguments, its given to them when the render method is called.

Leguan16 replied:
I do that in the next section.

Machine-Maker reviewed (changes requested):
This should also mention the use of ChatRenderer.viewerUnaware as something that should be used if the message will not be different depending on who is getting it. This is more performant because the message will only be rendered once per player it's being sent to.

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,156 @@
+    public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
+        User user = LuckPermsProvider.get().getPlayerAdapter(Player.class).getUser(source);
+        String prefixString = user.getCachedData().getMetaData().getPrefix();
+        Component prefix = MiniMessage.miniMessage().deserialize(prefixString);
```
This will throw if the prefix is null, also if we're going to use LuckPerms as an example, might want to link to the api docs for the LP API?

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,156 @@
+## Conclusion
+
+That is all you need to know about the new chat event and its renderer. 
+Of course there are many more things you can do with Component in general. 
```
```suggestion
Of course there are many more things you can do with Components in general. 
```

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,156 @@
+
+- /lp creategroup admin
+
+- /lp group admin meta addprefix 100 "`<gray>[</gray><red>Admin</red><gray>]</gray>`"
```
Why is the prefix in backticks here?

Leguan16 replied:
without backticks it produces this: 
![image](https://github.com/PaperMC/docs/assets/59799222/82865e2b-3692-465f-80e8-22c4f2702a92)

powercasgamer replied:
Put the whole command in backticks?

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,156 @@
+---
+slug: /dev/chat-events
+description: How to properly use the new AsyncChatEvent 
```
personally not a big fan of this description

olijeffers0n replied:
```suggestion
description: An outline on the AsyncChatEvent and how to handle it
```

olijeffers0n commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,156 @@
+---
+
+# Chat Events
+The chat event has evolved a couple of times over the years. 
```
```suggestion
The Chat event has evolved a few times over the years. 
```

Leguan16 replied:
Why exactly should "Chat" be uppercase here?

olijeffers0n replied:
Looks nicer i think… doesn’t need to be though

olijeffers0n commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,156 @@
+## Understanding the Renderer
+
+Before we can start using the new chat event, we need to understand how the new renderer works.
+The renderer is Paper's way of allowing plugins to modify the chat message before it is sent to the player.
```
Paper or Adventure?

Leguan16 replied:
The `ChatRenderer` class is located in `io.papermc.paper.chat.ChatRenderer` so I would say it's papers way of modifying chat messages.

olijeffers0n commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,156 @@
+::: info
+
+`MiniMessage` is a library that allows you to serialize and deserialize `Component` to and from a string using 
+readable html tags. You can read more about it [here](https://docs.advntr.dev/minimessage.html).
```
```suggestion
readable XML tags. You can read more about it [here](https://docs.advntr.dev/minimessage.html).
```

olijeffers0n commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,156 @@
+
+That is all you need to know about the new chat event and its renderer. 
+Of course there are many more things you can do with Component in general. 
+If you want to learn more about Component, you can read the [Component Documentation](https://docs.advntr.dev/text.html).
```
```suggestion
If you want to learn more about Components, you can read the [Component Documentation](https://docs.advntr.dev/text.html).
```

olijeffers0n reviewed (changes requested):
Maybe some comments on the fact that it is Asynchronous as well? Looks like a good addition though

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
````diff
@@ -0,0 +1,156 @@
+
+Another way of using the renderer is by using a lambda expression.
+
+```java title="ChatEvent.java"
+public class ChatEvent implements Listener {
````
```suggestion
```java title="ChatListener.java"
public class ChatListener implements Listener {
```

Leguan16 commented:
I addressed the requested changes. 
Feedback about the placement of the docs/wiki link of LuckPerms and MiniMessage would be appreciated.

kezz commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,152 @@
+
+    @EventHandler
+    public void onChat(AsyncChatEvent event) {
+        event.renderer(new ExampleRenderer());
```
Surely it's worth mentioning that you only need to instantiate this once? Or have `ChatListener` implement `ChatRenderer` and just pass `this`? Idk why we'd recommend instantiating a new object every time for no reason

Leguan16 replied:
good point

kezz commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,152 @@
+
+Now you can see that the message is rendered as we wanted it to be.
+
+## Rendering the message with a prefix using LuckPerms and MiniMessage
```
don't really see why this is useful:
1) plugins already exist that do this including one shipped by LP
2) nobody would realistically be hardcoding the format like that, it should be using placeholders instead
3) now we have to maintain docs based on an entirely external plugin

would be better to make it generic, just having a "how to use a configurable format" and then just have the example be MiniMessage with placeholders and a little "add your own here!" bit or something

kezz commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,152 @@
+
+Before we can start using the new chat event, we need to understand how the new renderer works.
+The renderer is Paper's way of allowing plugins to modify the chat message before it is sent to the player.
+This is done by using the `ChatRenderer` interface with its `render` method. Previously, this was done by using the `AsyncPlayerChatEvent` with its `setFormat` method.
```
don't know why we'd bother mentioning the deprecated method here

Leguan16 replied:
because it was requested:
https://github.com/PaperMC/docs/pull/264#discussion_r1382642954

Machine-Maker replied:
@kezz I think it's good to mention what it replaces. We frequently get questions about what the replacement for setFormat is on discord, enough that I think its warranted mentioning it somewhere, somehow.

Leguan16 commented:
I added a note regarding the event being async, added a tip about the usage of `ChatRenderer.ViewerUnaware` and addressed kezz's comment about the renderer being instantiated every time the event is called. 

I did not yet change the last section.

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
````diff
@@ -0,0 +1,173 @@
+In this example, we will be using our ChatListener class. 
+Next, we need to tell the event to use the renderer by using the `renderer` method.
+
+```java title="ChatListner.java"
+public class ChatListner implements Listener, ChatRenderer { // Implement the ChatRenderer and Listener interface
````
```suggestion
```java title="ChatListener.java"
public class ChatListener implements Listener, ChatRenderer { // Implement the ChatRenderer and Listener interface
```

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
````diff
@@ -0,0 +1,173 @@
+
+Another way of using the renderer is by using a lambda expression.
+
+```java title="ChatListner.java"
+public class ChatListner implements Listener {
````
```suggestion
```java title="ChatListener.java"
public class ChatListener implements Listener {
```

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
````diff
@@ -0,0 +1,173 @@
+
+To do this, we need to return a new `Component` that contains the message we want to send.
+
+```java title="ChatListner.java"
+public class ChatListner implements Listener, ChatRenderer {
````
```suggestion
```java title="ChatListener.java"
public class ChatListener implements Listener, ChatRenderer {
```

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,173 @@
+    @Override
+    public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
+        User user = LuckPermsProvider.get().getPlayerAdapter(Player.class).getUser(source);
+        String prefixString = user.getCachedData().getMetaData().getPrefix();
+        Component prefix = MiniMessage.miniMessage().deserialize(prefixString);
```
null check

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,173 @@
+The key difference between the `AsyncChatEvent` and the `ChatEvent` is that the `AsyncChatEvent` is fired asynchronously 
+which means that it does not block the main thread and delays the chat message.
+Be aware that using the Bukkit API in the event might cause errors. 
+If you need to use the Bukkit API, use the `ChatEvent` instead.
```
"...use the BukkitScheduler or the `ChatEvent` instead" something like this

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,173 @@
+
+:::
+
+## Understanding the Renderer
```
```suggestion
## Understanding the renderer
```

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,173 @@
+
+Now that we understand how the renderer works, we can start using it.
+
+## Using the Renderer
```
```suggestion
## Using the renderer
```

powercasgamer commented on docs/paper/dev/api/event-api/chat-event.md:
````diff
@@ -0,0 +1,138 @@
+
+To do this, we need to return a new `Component` that contains the message we want to send.
+
+```java title="ChatListener.java"
+public class ChatListener implements Listener, ChatRenderer {
+
+    @Override
+    public @NotNull Component render(@NotNull Player source, @NotNull Component sourceDisplayName, @NotNull Component message, @NotNull Audience viewer) {
+        return sourceDisplayName 
+                .append(Component.text(": ")) 
+                .append(message); 
+    }
+
+}
+```
````
Missing the listener here?

Leguan16 replied:
Intended since it didn't change and would just distract from the changes in `render`

olijeffers0n commented on docs/paper/dev/api/event-api/chat-event.md:
```diff
@@ -0,0 +1,138 @@
+- The `source` parameter is the player that sent the message.
+- The `sourceDisplayName` parameter is the display name of the player that sent the message.
+- The `message` parameter is the message that was sent.
+- The `viewer` parameter is the player that is receiving the message.
```
am i right in saying this is often not a singular player but instead can be a collection of players?

Leguan16 replied:
Well in terms of this method it's a singular audience.

olijeffers0n reviewed (approved):
![image](https://github.com/PaperMC/docs/assets/69084614/a2c5e330-c5cd-4e64-8913-88b7f75a2359)
I think that the styling on some of the links is a bit wonkey, but looks generally good. Thats obviously not your fault
